### PR TITLE
feat: add optional logging for expander stages

### PR DESCRIPTION
**Pull request by xiaoweim**
This PR adds the option to enable logging for each expander stage for better debug ability.

Log message start with `expanderDebugLog---${Kind}/${Namespace}/${Name}---version: ${Version}`

Example log:
```
2024-06-29T00:35:34Z	DEBUG	events	expanderDebugLog---PConfig/team-a/team-a-config---version: 1	{"type": "Normal", "object": {"kind":"PConfig","namespace":"team-a","name":"team-a-config","uid":"9d674656-1db3-4d54-b5eb-f51bf90d8732","apiVersion":"facade.foocorp.com/v1alpha1","resourceVersion":"679"}, "reason": "Running expander stage 0: common"}
```

```
2024-06-29T00:35:34Z	DEBUG	events	expanderDebugLog---PConfig/team-a/team-a-config---version: 1---request: config:"apiVersion: v1\nkind: ConfigMap\nmetadata:\n  name: common-config\n  namespace: {{ pconfigs.metadata.namespace }}\n  labels:\n    createdby: \"composition-namespaceconfigmap\"\ndata:\n  key1: value1" context:"{\"apiVersion\":\"composition.google.com/v1alpha1\",\"kind\":\"Context\",\"metadata\":{\"creationTimestamp\":\"2024-06-29T00:35:33Z\",\"generation\":1,\"managedFields\":[{\"apiVersion\":\"composition.google.com/v1alpha1\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:spec\":{\".\":{},\"f:project\":{}}},\"manager\":\"testcases.test\",\"operation\":\"Update\",\"time\":\"2024-06-29T00:35:33Z\"}],\"name\":\"context\",\"namespace\":\"team-a\",\"resourceVersion\":\"678\",\"uid\":\"aafd1a60-c07c-4d2e-95e7-059be18ef8b9\"},\"spec\":{\"project\":\"proj-a\"}}" facade:"{\"apiVersion\":\"facade.foocorp.com/v1alpha1\",\"kind\":\"PConfig\",\"metadata\":{\"annotations\":{\"composition-expander-debug-logs\":\"true\"},\"creationTimestamp\":\"2024-06-29T00:35:33Z\",\"generation\":1,\"managedFields\":[{\"apiVersion\":\"facade.foocorp.com/v1alpha1\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:composition-expander-debug-logs\":{}}},\"f:spec\":{\".\":{},\"f:project\":{}}},\"manager\":\"testcases.test\",\"operation\":\"Update\",\"time\":\"2024-06-29T00:35:33Z\"}],\"name\":\"team-a-config\",\"namespace\":\"team-a\",\"resourceVersion\":\"679\",\"uid\":\"9d674656-1db3-4d54-b5eb-f51bf90d8732\"},\"spec\":{\"project\":\"proj-a\"}}" value:"{}" resource:"pconfigs", result: error:{} type:MANIFESTS manifests:"apiVersion: v1\nkind: ConfigMap\nmetadata:\n  name: common-config\n  namespace: team-a\n  labels:\n    createdby: \"composition-namespaceconfigmap\"\ndata:\n  key1: value1"	{"type": "Normal", "object": {"kind":"PConfig","namespace":"team-a","name":"team-a-config","uid":"9d674656-1db3-4d54-b5eb-f51bf90d8732","apiVersion":"facade.foocorp.com/v1alpha1","resourceVersion":"679"}, "reason": "Expander stage common evaluation completed"}
```

```
2024-06-29T01:02:23Z	DEBUG	events	expanderDebugLog---PConfig/team-a/team-a-config---version: 1---resource count: 1	{"type": "Normal", "object": {"kind":"PConfig","namespace":"team-a","name":"team-a-config","uid":"431e53bf-1c9c-4ee2-873d-bdf4028c98db","apiVersion":"facade.foocorp.com/v1alpha1","resourceVersion":"678"}, "reason": "Finished expander stage 0: common"}
```

### CODE REVIEW

File: experiments/compositions/composition/internal/controller/expander_reconciler.go [hunk 1]
```diff
@@ -221,6 +222,17 @@ func (r *ExpanderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 
 	// ------------------- APPLIER SECTION -----------------------
 	stagesApplied := []string{}
+	expanderLogsEnabled := false
+	var err error
+	value, exist := inputcr.GetAnnotations()["composition-expander-stage-logs"]
+	if !exist {
+		logger.Info("annotation does not exist.")
```
Comment (barney-s) [inline experiments/compositions/composition/internal/controller/expander_reconciler.go:229]: log what annotation doesnt exist.

Comment (barney-s) [reply experiments/compositions/composition/internal/controller/expander_reconciler.go:229]: I would log if the annotation exists and not log if it doesnt exist (more common use case)

File: experiments/compositions/composition/tests/data/TestSimpleCompositionExpanderLoggingEnabled/input.yaml [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+spec:
+  project: proj-a
+---
+apiVersion: facade.foocorp.com/v1alpha1
+kind: PConfig
+metadata:
+  name: team-a-config
+  namespace: team-a
+  annotations:
+    composition-expander-stage-logs: "true"
```
Comment (barney-s) [inline experiments/compositions/composition/tests/data/TestSimpleCompositionExpanderLoggingEnabled/input.yaml:159]: composition-expander-debug-logs ?

Comment (xiaoweim) [reply experiments/compositions/composition/tests/data/TestSimpleCompositionExpanderLoggingEnabled/input.yaml]: Ack.

File: experiments/compositions/composition/internal/controller/expander_reconciler.go (hunk 1 above)
Comment (xiaoweim) [reply experiments/compositions/composition/internal/controller/expander_reconciler.go]: Make sense.

File: experiments/compositions/composition/internal/controller/expander_reconciler.go [hunk 3]
```diff
@@ -324,6 +331,16 @@ func (r *ExpanderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		newStatus.ClearCondition(compositionv1alpha1.Ready)
 		message := fmt.Sprintf("Evaluated stages: %s \n Applied stages: %s", strings.Join(stagesEvaluated, ", "), strings.Join(stagesApplied, ", "))
 		newStatus.AppendCondition(compositionv1alpha1.Ready, metav1.ConditionFalse, message, "PendingStages")
+
+		if expanderLogsEnabled {
+			// TODO(@xiaoweim): Log ressource UID in the future if possible
+			r.Recorder.Event(&inputcr, "Normal", fmt.Sprintf("Finished expander stage %d: %s", index, expander), fmt.Sprintf("version: %d, resource count: %d", inputcr.GetGeneration(), applier.Count()))
+			for i, resourceStatus := range newStatus.Stages[expander].LastApplied {
+				logger.Info("Resource %d\n", i)
+				logger.Info("Resource name: %s, resource namespace: %s, resource gvk: %s %s %s, resource status: %s",
```
Comment (barney-s) [inline experiments/compositions/composition/internal/controller/expander_reconciler.go:340]: I could be wrong. But i think @cheftako's intent was to include the whole object here. 
I would do it evaluateAndSavePlan() where we call the GRPC interface.

The goal is to record the GRPC inputs and response.

File: experiments/compositions/composition/internal/controller/expander_reconciler.go [hunk 4]
```diff
@@ -324,6 +335,16 @@ func (r *ExpanderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		newStatus.ClearCondition(compositionv1alpha1.Ready)
 		message := fmt.Sprintf("Evaluated stages: %s \n Applied stages: %s", strings.Join(stagesEvaluated, ", "), strings.Join(stagesApplied, ", "))
 		newStatus.AppendCondition(compositionv1alpha1.Ready, metav1.ConditionFalse, message, "PendingStages")
+
+		if expanderDebugLogsEnabled {
+			// TODO(@xiaoweim): Log ressource UID in the future if possible
+			r.Recorder.Event(&inputcr, "Normal", fmt.Sprintf("Finished expander stage %d: %s", index, expander), fmt.Sprintf("version: %d, resource count: %d", inputcr.GetGeneration(), applier.Count()))
+			for i, resourceStatus := range newStatus.Stages[expander].LastApplied {
+				logger.Info("Resource %d\n", i)
```
Comment (cheftako) [inline experiments/compositions/composition/internal/controller/expander_reconciler.go:343]: Why is this line separate? There is no guaranteed way to associate it with the output on line 344 once the system is busy.

File: experiments/compositions/composition/internal/controller/expander_reconciler.go [hunk 5]
```diff
@@ -324,6 +335,16 @@ func (r *ExpanderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
+				logger.Info("Resource name: %s, resource namespace: %s, resource gvk: %s %s %s, resource status: %s",
```
Comment (cheftako) [inline experiments/compositions/composition/internal/controller/expander_reconciler.go:344]: I would consider having a method which returned an identifier string and call it at the start of each expander debug log line. You then know you are printing the same identifying log id everywhere. You even know the same grep will work for all of them.

File: experiments/compositions/composition/internal/controller/expander_reconciler.go [hunk 6]
```diff
@@ -517,6 +540,10 @@ func (r *ExpanderReconciler) evaluateAndSavePlan(ctx context.Context, logger log
 		err = fmt.Errorf("Evaluate Failed: %s", result.Error.Message)
 		return values, updated, "EvaluateStatusFailed", err
 	}
+	if expanderDebugLogEnabled {
+		logger.Info(expanderDebugLog(cr) + fmt.Sprintf("---sent expander request: %v, received results: %v", evaluateRequest, result))
+		r.Recorder.Event(cr, "Normal", fmt.Sprintf("Expander stage %s evaluation completed", expander.Name), expanderDebugLog(cr)+fmt.Sprintf("---request: %v, result: %v", evaluateRequest, result))
```
Comment (barney-s) [inline experiments/compositions/composition/internal/controller/expander_reconciler.go:545]: Not sure if there is a limit on what can be sent to an event. Nevertheless lets try it.

**Review by barney-s — APPROVED** [review]
/approve
/lgtm

**Pull request merged**